src/bc/de.rs

```diff
@@ -116,99 +116,100 @@ fn bc_modern_msg<'a, 'b>(
+                passwd: None,
+            })),
+            _ => context.set_encrypted(EncryptionProtocol::Unknown),
         }
-    };
+    }
 
-    let (mut buf, body_buf) = take(end_of_xml)(buf)?;
+    let (buf, ext_buf) = take(ext_len)(buf)?;
+    let payload_len = header.body_len - ext_len;
+    let (buf, payload_buf) = take(payload_len)(buf)?;
 
     let decrypted;
-    let processed_body_buf = if !header.is_encrypted() {
-        buf
-    } else {
-        decrypted = xml_crypto::crypt(header.enc_offset, body_buf);
-        &decrypted
+    let processed_ext_buf = match context.get_encrypted() {
+        EncryptionProtocol::Unencrypted => ext_buf,
+        encryption_protocol => {
+            decrypted = xml_crypto::crypt(header.channel_id as u32, ext_buf, encryption_protocol);
+            &decrypted
+        }
     };
 
     // Now we'll take the buffer that Nom gave a ref to and parse it.
-    let mut xml = None;
-    let mut binary = None;
-    if end_of_xml > 0 {
+    let extension;
+    if ext_len > 0 {
         // Apply the XML parse function, but throw away the reference to decrypted in the Ok and
         // Err case. This error-error-error thing is the same idiom Nom uses internally.
-        let parsed = AllTopXmls::try_parse(processed_body_buf)
+        let parsed = Extension::try_parse(processed_ext_buf)
             .map_err(|_| Err::Error(make_error(buf, ErrorKind::MapRes)))?;
-
-        // If this is the first message containing binary, the Extension message puts the message
-        // ID into binary mode, then the first binary is sent after the XML.  All remaining
-        // messages for that ID are pure binary.
-        match parsed {
-            AllTopXmls::BcXml(x) => {
-                xml = Some(x);
-            }
-            AllTopXmls::Extension(Extension { binary_data: _ }) => {
-                in_bin_mode = true;
-            }
-        }
+        extension = Some(parsed);
+    } else {
+        extension = None;
     }
 
-    // If we are in binary mode, extract it
-    if in_bin_mode {
-        if let Some(bin_offset) = header.bin_offset {
-            // Extract remainder of message as binary, if it exists
-            let (buf_after, payload) =
-                map(take(header.body_len - bin_offset), |x: &[u8]| x.to_vec())(buf)?;
-
-            // Since the parser operates in streaming mode, must wait until after we successfully
-            // receive enough bytes before modifying the context (otherwise we'll alter the
-            // behavior of future passes of this function even if we didn't yet consume the
-            // message).
-            context.in_bin_mode.insert(header.msg_id);
-
-            binary = Some(payload);
-            buf = buf_after;
+    // Now to handle the payload block
+    // This block can either be xml or binary depending on what the message expects.
+    // For our purposes we use try_parse and if all xml based parsers fail we treat
+    // As binary
```

Comment:
Although, I don't really understand the Extension vs Binary distinction still (maybe not enough time spent staring at wireshark). Could you explain the packet format again?

Reply:
You remember the binary_offset number.

Everything before that offset is 

Extension

Everything after that offset is either

Binary or XML.

The extension can tell you if it's binary or XML. But you have to cache it. Per message_id (not per message type). Because the sever can send extension then payload in seperate packets I.e. video. So I could make it know prior by storing the context with a hashmap.

This is why I renamed binary offset to payload offset. Because it's only binary sometimes.

Reply:
You can have two seperate XML blocks in one message seperated by the payload_offset. Extension XML and BC XML. Which is the part needed for most command and control requests including motion detection and the mqtt bit in the other PR.

Reply:
The last update we made to the dissector includes these changes for extension and payload xmls.

Reply:
> Everything before that offset is Extension XML. Everything after that offset is either Binary or BC XML.

Ahhh got it thank you. Yes, this makes sense. But I think I would still prefer to cache whether the payload is binary.

Reply:
Rather than cache just binary or not should we cache the whole extension xml? This xml is supposed to tell use how to decode the payload to follow. If it cache it per message number though this cache could get quite large.

Reply:
Ideally I'd like to move this into the subscriber somehow. A single message num corresponds to a round trip. client->camera->client. Currently we subscribe based on message type 1->login, 3->video etc but I think it would be better to subscribe on the message num to ensure the data corresponds to your request rather than a race with something else requesting the same message type. This subscribed object should also keep the extension info it needs and when we close the subscribe we can safely discard the extension

Reply:
For now I will just use the context. But its something to think about

Reply:
Done with context and message number